core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -687,6 +687,9 @@ private[spark] class AppStatusListener(
         stage.killedSummary = killedTasksSummary(event.reason, stage.killedSummary)
       }
       stage.activeTasksPerExecutor(event.taskInfo.executorId) -= 1
+
+      stage.executorSummary(event.taskInfo.executorId).peakExecutorMetrics
+        .compareAndUpdatePeakValues(event.taskExecutorMetrics)
```

Comment:
Hmm... executor metrics for each stage should be collected [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/AppStatusListener.scala#L980).
But if the heartbeat interval from an executor is longer than lifetime of a stage, we can't collect the executor metrics.
So this change can be one option. What do you think @gengliangwang ?

Reply:
We collect all data to choose the peak metrics. IMO,  more accuracy is better. 
But I'm not particularly clear about the performance impact of this part, hope more suggestion.

Reply:
@imback82 Do you have any concern about this change?

Reply:
This seems fine to me if we need more accurate peak values.

Reply:
It seems that the first metrics of `peakExecutorMetrics` become 0 instead of -1 after this. @AngersZhuuuu Do you know the reason?

Reply:
> instead

`-1` is default peakExecutorMetrics, and with this change,  we will update this value with real task metrics, but why all is `0` depend on the metrics data...

Reply:
> It seems that the first metrics of peakExecutorMetrics become 0 instead of -1 after this. @AngersZhuuuu Do you know the reason?

@gengliangwang 
By this change, `peakExecutorMetrics` is updated not only `onExecutorMetricsUpdate` but also `onTaskEnd`.
So, the peak value carried by `SparkListenerTaskEnd` is `0`, the corresponding peak values in `peakExecutorMetrics` is set to `0`.
Do you have any concern?

Reply:
@sarutak No, I am ok with it :)

Reply:
O.K, I'll merge later if there are no objections. Thanks for the response. @gengliangwang